feat: testing pagina voor developers (#3101)
nl-design-system/documentatie: docs/handboek/developer/15-component-testen.md

---
title: Componenten testen · Developer · Handboek
hide_title: true
hide_table_of_contents: false
sidebar_label: Componenten testen
pagination_label: Componenten testen
description: Eigen componenten testen
keywords:
  - developer
  - bijdragen
  - componenten testen
---

# Componenten testen

Op deze pagina vindt je tips om componenten te testen op de juiste aspecten.

## Testen op uitbreidbaarheid

> Zoals op de [bijdragen pagina](/handboek/developer/bijdragen) is beschreven, is uitbreidbaarheid erg belangrijk. Daarom is het van belang om uitbreidbaarheid te testen.

Reviewer: Deze zin is dubbel: het is belangrijk om te testen omdat het belangrijk is

### Class-namen

> Front-end developers vertrouwen op de BEM class-namen om hun eigen CSS toe te voegen. Wanneer de component een class-naam hernoemt of verwijdert, is er sprake van een 'breaking change'. Unit tests moeten elke class-naam controleren, zodat het betrouwbare API's zijn. Je zult veel tests zoals deze tegenkomen:

Reviewer: Ik zie in 03-thema-maken dat we wel "classname" gebruiken, misschien moeten we dat hier doen
```suggestion
Front-end developers vertrouwen op de BEM classnames om hun eigen CSS toe te voegen. Wanneer de component een classname hernoemt of verwijdert, is er sprake van een 'breaking change'. Unit tests moeten elke classname controleren, zodat het betrouwbare API's zijn. Je zult veel tests zoals deze tegenkomen:
```

  const { container } = render(<MyComponent />);

  const field = container.querySelector("div");

  expect(field).toHaveClass(".my-component");
});
```

## HTML-content in componenten

Tekst in componenten kan soms worden verbeterd met markup: taalmetadata, code, nadruk of afbeeldingen. Elke property die in de HTML terechtkomt, moet worden getest op uitbreidbaarheid met rich text-content.

```javascript
it("renders rich text content", () => {
  const { container } = render(
    <Heading1 {...defaultProps}>
      The French national motto: <span lang="fr">Liberté, égalité, fraternité</span>
    </Heading1>,
  );

  const richText = container.querySelector("span");

  expect(richText).toBeInTheDocument();
});
```

Het testen van properties is misschien nog wel belangrijker, omdat children meestal al HTML-content toestaat:

```javascript
it('renders rich text content', () => {
  const { container } = render(
    <FormFieldTextbox label={
      <EmailIcon/> E-mail address
    }></FormFieldTextbox>,
  );

  const richText = container.querySelector('svg');

  expect(richText).toBeInTheDocument();
});
```

## Maak native HTML niet kapot

### Globale attributen

> Globale attributen kunnen op alle HTML-elementen worden gebruikt, dus componenten die HTML renderen moeten deze ook ondersteunen. In React is dit eenvoudig te ondersteunen met ...restProps. De volgende codevoorbeelden gebruiken globale attributen:

Reviewer: Kan dit misschien iets duidelijker, wat restProps precies is, inclusief een voorbeeld hoe je de restProps instelt?

Bijvoorbeeld in de trant van:

---

Zoals aangegeven in [de API conventie voor componenten voor HTML](https://nldesignsystem.nl/handboek/developer/api-conventie/#componenten-voor-html) moeten [Globale HTML-attributen](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes) hun normale gedrag behouden. In React kun je alle overige, niet-gespecificeerde properties opvangen met de zogenaamde rest-operator (meestal door `..restProps` te spreaden):

> Ik heb deze tekst vooral gekopieerd uit de API conventie, maar het is wel een goede tekst.

> Willen we hier linken naar https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters of doen we dat niet?

```jsx
({ someProp, ...restProps })
```

De `someProp` wordt expliciet uit het object gehaald, terwijl `restProps` een object bevat met alle overige properties die aan het component zijn doorgegeven.

Daardoor kun je bijvoorbeeld alle globale HTML-attributen automatisch doorgeven aan het onderliggende HTML-element:

```jsx
const MyComponent = ({ someProp, ...restProps }) => {
  return <div {...restProps}>{someProp}</div>;
};
```

Zo worden alle HTML-attributen (zoals `id`, `className`, `data-*`, `aria-*`, etc.) die je op `<MyComponent>` zet automatisch op de `<div>` toegepast.

```XML
<MyComponent id="main" />
<MyComponent style={{ '--my-component-color': 'currentColor' }} />
<MyComponent hidden />
<MyComponent tabIndex={-1} />
<MyComponent lang="en" />
<MyComponent className="custom" />
<MyComponent data-test-id="component" />
<MyComponent role="group" />
```

Author: Thanks voor de suggestie!

> Willen we hier linken naar https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters of doen we dat niet?

Hoeft denk ik niet, misschien kan het wat verwarrend overkomen en mensen googlen al vrij snel naar informatie op MDN als ze iets zoeken denk ik.

```XML
<MyComponent id="main" />
<MyComponent style={{ '--my-component-color': 'currentColor' }} />
<MyComponent hidden />
<MyComponent tabIndex={-1} />
<MyComponent lang="en" />
<MyComponent className="custom" />
<MyComponent data-test-id="component" />
<MyComponent role="group" />
```

### Het hidden attribuut

> De CSS voor een component breekt vaak het hidden attribuut, omdat code zoals display: flex de standaardstijlen overschrijft. Test of het hidden attribuut het [element] nog steeds onzichtbaar maakt.

Reviewer 2: ```suggestion
De CSS voor een component breekt vaak het `hidden`-attribuut, omdat code zoals `display: flex` de standaardstijlen overschrijft. Test of het `hidden`-attribuut het [element] nog steeds onzichtbaar maakt.
```

> De CSS voor een component breekt vaak het `hidden`-attribuut, omdat code zoals `display: flex` de standaardstijlen overschrijft. Test of het `hidden`-attribuut het [element] nog steeds onzichtbaar maakt.

Reviewer: ```suggestion
De CSS voor een component breekt vaak het `hidden`-attribuut, omdat code zoals `display: flex` de standaardstijlen overschrijft. Test of het `hidden`-attribuut het HTML-element nog steeds onzichtbaar maakt.
```
Waarom had je [element]?

```javascript
it("can be hidden", () => {
  const { container } = render(<MyComponent hidden />);

  const component = container.querySelector("div");

  expect(component).not.toBeVisible();
});
```

## De className property

Componenten renderen BEM class-namen, maar front-end developers moeten ook hun eigen class-namen kunnen gebruiken. Extra class-namen moeten de lijst van class-namen uitbreiden, niet de class-namen van de component overschrijven.

> ```javascript

Reviewer: Willen we hier ook meteen laten zien hoe je dit voor je code kunt doen? Dus een component voorbeeld waar de classnames worden gecombineerd? Of is dat out of scope voor deze pagina?

it("can have a additional class name", () => {
  const { container } = render(<MyComponent className="large" />);

  const component = container.querySelector(":only-child");

  expect(component).toHaveClass("large");
  expect(component).toHaveClass("my-component");
});
```

## Test de accessibility tree

### Landmarks

> ```javascript

Reviewer: Ik mis hier een uitleg of een korte introductie

---
title: Componenten testen · Developer · Handboek
hide_title: true
hide_table_of_contents: false
sidebar_label: Componenten testen
pagination_label: Componenten testen
description: Eigen componenten testen
keywords:
  - developer
  - bijdragen
  - componenten testen
  - css component testen
  - react.js component testen
---

# Componenten testen

Op deze pagina vind je tips om componenten te testen op de juiste aspecten. De meeste voorbeelden zijn uitgewerkt voor React.js met React Testing Library, maar kunnen soms op andere frameworks toegepast worden.

## Testen op uitbreidbaarheid

> Zoals op de [bijdragen pagina](/handboek/developer/bijdragen) is beschreven, is uitbreidbaarheid erg belangrijk. Daarom is het van belang om dit te testen.

Reviewer: ```suggestion
Zoals op de [bijdragen pagina](/handboek/developer/bijdragen) staat, is uitbreidbaarheid belangrijk. Hieronder lees je hoe je dit test.
```

Om het dubbele, belangrijk = belangrijk te fixen.
